[IsDeprecatedWeakRefSmartPointerException] Make SpeechSynthesisClient ref-counted

RupinMittal opened the pull request:
#### 4e1ba66c57dfdc05d16b3e221ca567d5da4f000e
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make SpeechSynthesisClient ref-counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=280555">https://bugs.webkit.org/show_bug.cgi?id=280555</a>
<a href="https://rdar.apple.com/136865003">rdar://136865003</a>

Reviewed by Chris Dumez.

We remove IsDeprecatedWeakRefSmartPointerException by making SpeechSynthesisClient
ref-counted.

WebSpeechSynthesisClient is owned by Page. Since it&apos;s &apos;now ref-counted, it&apos;s possible
for it to out-live the Page, which would be a problem. To ensure that
WebSpeecSynthesisClient never tries to use it&apos;s WebPage member if the Page has been
destroyed, we change the m_page member to be a WeakPtr instead of a WeakRef and
null-check m_page before using it.

* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::create):
(WebCore::SpeechSynthesis::SpeechSynthesis):
(WebCore::SpeechSynthesis::getVoices):
(WebCore::SpeechSynthesis::startSpeakingImmediately):
(WebCore::SpeechSynthesis::cancel):
(WebCore::SpeechSynthesis::pause):
(WebCore::SpeechSynthesis::resumeSynthesis):
(WebCore::Ref&lt;SpeechSynthesis&gt;SpeechSynthesis::create): Deleted.
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/SpeechSynthesisClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.cpp:
(WebKit::WebSpeechSynthesisClient::voiceList):
(WebKit::WebSpeechSynthesisClient::corePageObserver const):
(WebKit::WebSpeechSynthesisClient::resetState):
(WebKit::WebSpeechSynthesisClient::speak):
(WebKit::WebSpeechSynthesisClient::cancel):
(WebKit::WebSpeechSynthesisClient::pause):
(WebKit::WebSpeechSynthesisClient::resume):
(WebKit::WebSpeechSynthesisClient::protectedPage const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.h:
(WebKit::WebSpeechSynthesisClient::~WebSpeechSynthesisClient): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: <a href="https://commits.webkit.org/284434@main">https://commits.webkit.org/284434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c6ecbe1a29f3741b81eb749d28deca50fdf5fb8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48699 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73381 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20457 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55113 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13577 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72365 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44434 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59830 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35592 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17260 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18833 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63042 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75093 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13281 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62779 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13320 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59913 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62686 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15392 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10703 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4314 "Passed tests") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44503 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45577 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46772 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.cpp:
```diff
@@ -118,7 +128,8 @@ void WebSpeechSynthesisClient::pause()
     };
 
-    protectedPage()->sendWithAsyncReply(Messages::WebPageProxy::SpeechSynthesisPause(), WTFMove(completionHandler));
+    if (RefPtr page = m_page.get())
```
You're failing to call the completionHandler in the `else` case, which will cause crashes.

RupinMittal replied:
Maybe a dumb question--why would this cause crashes? Won't the completion handler just go out of scope if we don't call it? Also, does this new version work? (Null-check page and return before the completion handler is declared instead of an else clause that calls the completion handler).

cdumez replied:
CompletionHandler has an assertion in its destructor to make sure that it has been called. Not calling a completion handler will generally lead to hangs or breakage so the assertion makes sure we catch these cases.

RupinMittal replied:
Ahhh ok that makes sense. This new iteration of the code null-check page and returns before the completion handler is created if page is null. So we won't hit the assertion now.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.cpp:
```diff
@@ -130,7 +141,8 @@ void WebSpeechSynthesisClient::resume()
     };
 
-    protectedPage()->sendWithAsyncReply(Messages::WebPageProxy::SpeechSynthesisResume(), WTFMove(completionHandler));
+    if (RefPtr page = m_page.get())
```
ditto.

RupinMittal replied:
Updated the code to null-check page and return if it's null before the completion handler is created.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.h:
```diff
@@ -30,43 +30,42 @@
-    
-class WebSpeechSynthesisClient : public WebCore::SpeechSynthesisClient {
+
+class WebSpeechSynthesisClient : public RefCounted<WebSpeechSynthesisClient>, public WebCore::SpeechSynthesisClient {
```
Cam we mark the class as `final`?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2901,12 +2901,12 @@ void WebPage::setSuppressScrollbarAnimations(bool suppressAnimations)
 {
     m_page->setShouldSuppressScrollbarAnimations(suppressAnimations);
 }
-    
```
Please revert all these spacing changes and update your IDE configuration to not do this.

RupinMittal replied:
Sorry about that--was trying out a different editor.

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make SpeechSynthesisClient ref-counted
https://bugs.webkit.org/show_bug.cgi?id=280555
rdar://136865003

Reviewed by Chris Dumez.

We remove IsDeprecatedWeakRefSmartPointerException by making SpeechSynthesisClient
ref-counted.

WebSpeechSynthesisClient is owned by Page. Since it's 'now ref-counted, it's possible
for it to out-live the Page, which would be a problem. To ensure that
WebSpeecSynthesisClient never tries to use it's WebPage member if the Page has been
destroyed, we change the m_page member to be a WeakPtr instead of a WeakRef and
null-check m_page before using it.

* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::create):
(WebCore::SpeechSynthesis::SpeechSynthesis):
(WebCore::SpeechSynthesis::getVoices):
(WebCore::SpeechSynthesis::startSpeakingImmediately):
(WebCore::SpeechSynthesis::cancel):
(WebCore::SpeechSynthesis::pause):
(WebCore::SpeechSynthesis::resumeSynthesis):
(WebCore::Ref<SpeechSynthesis>SpeechSynthesis::create): Deleted.
* Source/WebCore/Modules/speech/SpeechSynthesis.h:
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/SpeechSynthesisClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.cpp:
(WebKit::WebSpeechSynthesisClient::voiceList):
(WebKit::WebSpeechSynthesisClient::corePageObserver const):
(WebKit::WebSpeechSynthesisClient::resetState):
(WebKit::WebSpeechSynthesisClient::speak):
(WebKit::WebSpeechSynthesisClient::cancel):
(WebKit::WebSpeechSynthesisClient::pause):
(WebKit::WebSpeechSynthesisClient::resume):
(WebKit::WebSpeechSynthesisClient::protectedPage const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebSpeechSynthesisClient.h:
(WebKit::WebSpeechSynthesisClient::~WebSpeechSynthesisClient): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: https://commits.webkit.org/284434@main
